### PR TITLE
Merge DBZ-175 into master

### PR DESCRIPTION
**Pull request by mtagle**
as discussed on the DBZ-175 ticket, merge the [DBZ-175](https://issues.jboss.org/browse/DBZ-175) feature branch into master.

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Thanks, @mtagle! Did you have a chance to do some more testing of this change in your environment?

**Comment by gunnarmorling** [comment]
@jpechane, I'd like to hear your opinion as well: do you agree to merging this eventually into master (amended by my fixes in #741)?

Please see [DBZ-1091](https://issues.jboss.org/browse/DBZ-1091) for follow-up tasks, esp. the discussion around leaking threads and unordered DB history during parallel log reading. Personally I think we can merge it (it's disabled by default), provided we follow up on the things touched on in DBZ-1091 quickly.

**Comment by mtagle** [comment]
@gunnarmorling Unfortunately we haven't gotten a chance to run this in a "real" environment yet. We've been having some trouble updating our environments to work with all the new upgraded versions of everything. Hopefully soon (as in, this month) though.

**Comment by gunnarmorling** [comment]
Ok. So what's your and @criccomini's take on this then, do you feel it's ready to go into next week's 0.9.0CR1? Also it would be great if you could review my changes added in #741. Thanks!

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/ReconcilingBinlogReader.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        unifiedReader.context.loadHistory(getLeadingReader().context.source());
+        unifiedReader.context.source().setFilterDataFromConfig(unifiedReader.context.config());
+        Map<String, ?> keyedOffset =
+            reconcilingReader.getLastOffset() == null ?
+                getLeadingReader().getLastOffset() :
+                reconcilingReader.getLastOffset();
+        unifiedReader.context.source()
+            .setBinlogStartPoint((String) keyedOffset.get(BINLOG_FILENAME_OFFSET_KEY),
+                                 (Long) keyedOffset.get(BINLOG_POSITION_OFFSET_KEY));
+        // note: this seems to dupe -one- event in my tests.
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/ReconcilingBinlogReader.java]: Uh, yes, that's a good point. In particular as I believe you're using GTIDs, right, @mtagle?

**Comment by gunnarmorling** [comment]
@jpechane, on the metrics one, can you clarify whether this just means metrics aren't avaible (which seems it could be fixed separately)?

**Comment by gunnarmorling** [comment]
@mtagle, as the commit history of this one is quite cluttered I'd like to squash all your commits into a single one, unless you have objections?

**Comment by mtagle** [comment]
I have absolutely no objections to squashing the commits.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/ReconcilingBinlogReader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtagle) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/ReconcilingBinlogReader.java]: As I understand it the GTIDs are handled by the .loadHistory() call on line 140. The .setBinlogStartPoint call only accepts a file and position, however.

**Comment by jpechane** [comment]
@mtagle IMHO GTID set is read from offsets and is set via `io.debezium.connector.mysql.SourceInfo.setCompletedGtidSet(String)`. I do not think it is related to schema history.

**Comment by gunnarmorling** [comment]
Merged the fix for the metrics name. Seems the GTID one is the last open issue for now.

**Comment by gunnarmorling** [comment]
Squashed all the initial commits into one and rebased. Waiting for CI and if it's green, I'll merge it.

**Comment by gunnarmorling** [comment]
Aaand it's merged to master 🎉!

Thanks a lot for your efforts and endurance, @mtagle! It would be fantastic if you could test this in your environment once we did the CR1 and before 0.9.0.Final is out.

For archeological purposes I've pushed the upstream/DBZ-175-pre-rebase-2019-01-14,
upstream/DBZ-175-pre-rebase-2018-10-08 and  upstream/DBZ-175-pre-rebase-2019-01-24 branches. I'll leave them there for some time, so we can dig into the detailed commit history of the feature, if needed.

I've logged [DBZ-1091](https://issues.jboss.org/browse/DBZ-1091) for remaining loose ends and follow-up tasks. I suggest to continue the discussion over there. Personally I'm planning to work on that suggested alternative parallel snapshotting mode, which will avoid the issues with leaking threads when shutting down the connector at the "wrong time". Another important task will be more complete integration tests around whitelist/blacklist config changes, so to avoid any regressions down the road.

In any case one take away for me is to plan for future changes to come in smaller chunks, so we can integrate them much quicker, without such long-running feature branch.

Moira, thanks again and looking forward to honing this feature together with you. We'll also have to look into rolling out the functionality to other connectors, but that's its own story :)

**Pull request merged**